app/models/product_importer.rb

```diff
@@ -177,7 +181,12 @@ def count_existing_products
           AND spree_variants.deleted_at IS NULL', supplier_id).
           count
 
-        @supplier_products[:by_supplier][supplier_id] = {existing_products: products_count}
+        if @supplier_products[:by_supplier][supplier_id]
+          @supplier_products[:by_supplier][supplier_id][:existing_products] = products_count
+        else
+          @supplier_products[:by_supplier][supplier_id] = {existing_products: products_count}
+        end
```

Comment:
Yep, I would create a `Product` or `Entry` class and map every row from the spreadsheet into an instance of that class.

What I'm suggesting is that if you use a more flexible object you can encapsulate logic and avoid relying on Hash `keys` presence. It would make all the workflow much more robust and explicit. But again, using Hash to just store numbers is OK. Just make sure that you don't need a better data structure to add logic and meaning to those numbers.